inst/include/cpp11/as.hpp

```diff
@@ -92,22 +92,21 @@ template <typename T>
 enable_if_integral<T, T> as_cpp(SEXP from) {
   if (Rf_isInteger(from)) {
     if (Rf_xlength(from) == 1) {
-      return INTEGER_ELT(from, 0);
+      return static_cast<T>(INTEGER_ELT(from, 0));
     }
   } else if (Rf_isReal(from)) {
     if (Rf_xlength(from) == 1) {
-      if (ISNA(REAL_ELT(from, 0))) {
-        return NA_INTEGER;
-      }
       double value = REAL_ELT(from, 0);
-      if (is_convertible_without_loss_to_integer(value)) {
-        return value;
+      if (ISNA(value) && sizeof(T) >= sizeof(int)) {
+        return static_cast<T>(NA_INTEGER);
+      } else if (!ISNA(value) && is_convertible_without_loss_to_integer(value)) {
+        return static_cast<T>(value);
       }
```

Comment:
This gave me pause. Perhaps:

```suggestion
      if (ISNA(value)) {
        if (sizeof(T) >= sizeof(int)) {
          return static_cast<T>(NA_INTEGER);
        }
        // For T of smaller size, can't do better than throw an error
      } else if (is_convertible_without_loss_to_integer(value)) {
        return static_cast<T>(value);
      }
```